build system: Rework EXTERNAL_MODULE_DIRS

maribu opened the pull request:
### Contribution description

Previously, external modules had to be individually added to both
EXTERNAL_MODULE_DIRS and USEMODULE. If those where not in sync, this
resulted in build errors.

With this commit, search folders for external modules are added to
EXTERNAL_MODULE_DIRS instead. So lets say the file system structure is
like this

```
└── /path/to/external/modules
    ├── mod_a
    │   ├── Makefile
    │   ├── Makefile.dep
    │   ├── Makefile.include
    │   ├── foo.c
    │   └── include
    │       └── external_module.h
    └── mod_b
        ├── Makefile
        └── bar.c

```

One now adds `/path/to/external/modules` to EXTERNAL_MODULES and only
with `USEMODULE += mod_a` the corresponding module, dependencies and
include settings are actually used. Hence, it is possible to configure
`EXTERNAL_MODULE_DIRS` from `~/.profile` or `~/.bashrc` once and never
needs to worry about them again.


### Testing procedure

The test application was updated to allow testing the new feature. But some packages also use this and have not been adapted. This is intended to be done if the general idea is approved.

### Issues/PRs references

Alternative to https://github.com/RIOT-OS/RIOT/pull/16103

gdoffe commented:
Seems really good @maribu . :)

Just a point, do you agree that makefiles/docker.inc.mk should be patched too to replace EXTERNAL_MODULE_DIRS by  EXTERNAL_MODULE_PATHS ?
```diff
gdo@gdo-desktop:~/Developpement/Informatique/Personnel/cogip/RIOT$ git diff ./makefiles/docker.inc.mk
diff --git a/makefiles/docker.inc.mk b/makefiles/docker.inc.mk
index 2e4e61f32f..2af7733743 100644
--- a/makefiles/docker.inc.mk
+++ b/makefiles/docker.inc.mk
@@ -253,7 +253,7 @@ DOCKER_VOLUMES_AND_ENV += $(if $(wildcard $(GIT_CACHE_DIR)),-e 'GIT_CACHE_DIR=$(
 
 # Remap external module directories.
 #
-# This remaps directories from EXTERNAL_MODULE_DIRS to subdirectories of
+# This remaps directories from EXTERNAL_MODULE_PATHS to subdirectories of
 # $(DOCKER_BUILD_ROOT)/external
 #
 # Remapped directories must all have different basenames
@@ -261,10 +261,10 @@ DOCKER_VOLUMES_AND_ENV += $(if $(wildcard $(GIT_CACHE_DIR)),-e 'GIT_CACHE_DIR=$(
 # Limitation: If a directory is inside RIOTPROJECT and not in RIOT it is
 # remapped anyway instead of loading from inside RIOTPROJECT.
 #
-# As EXTERNAL_MODULE_DIRS should ignore the 'Makefile' configuration, they must
+# As EXTERNAL_MODULE_PATHS should ignore the 'Makefile' configuration, they must
 # be set using command line variable settings to not be modified within docker.
-DOCKER_VOLUMES_AND_ENV += $(call docker_volumes_mapping,$(EXTERNAL_MODULE_DIRS),$(DOCKER_BUILD_ROOT)/external,)
-DOCKER_OVERRIDE_CMDLINE += $(call docker_cmdline_mapping,EXTERNAL_MODULE_DIRS,$(DOCKER_BUILD_ROOT)/external,)
+DOCKER_VOLUMES_AND_ENV += $(call docker_volumes_mapping,$(EXTERNAL_MODULE_PATHS),$(DOCKER_BUILD_ROOT)/external,)
+DOCKER_OVERRIDE_CMDLINE += $(call docker_cmdline_mapping,EXTERNAL_MODULE_PATHS,$(DOCKER_BUILD_ROOT)/external,)
 
 # Remap 'EXTERNAL_BOARD_DIRS' if they are external
 DOCKER_VOLUMES_AND_ENV += $(call docker_volumes_mapping,$(EXTERNAL_BOARD_DIRS),$(DOCKER_BUILD_ROOT)/external,)
@@ -275,13 +275,13 @@ DOCKER_OVERRIDE_CMDLINE += $(call docker_cmdline_mapping,EXTERNAL_BOARD_DIRS,$(D
 # External module directories sanity check:
 #
 # Detect if there are remapped directories with the same name as it is not handled.
-# Having EXTERNAL_MODULE_DIRS = /path/to/dir/name \
+# Having EXTERNAL_MODULE_PATHS = /path/to/dir/name \
 #                               /another/directory/also/called/name
 # would lead to both being mapped to '$(DOCKER_BUILD_ROOT)/external/name'
-_mounted_dirs = $(foreach d,$(EXTERNAL_MODULE_DIRS),$(if $(call dir_is_outside_riotbase,$(d)),$(d)))
+_mounted_dirs = $(foreach d,$(EXTERNAL_MODULE_PATHS),$(if $(call dir_is_outside_riotbase,$(d)),$(d)))
 ifneq ($(words $(sort $(notdir $(_mounted_dirs)))),$(words $(sort $(_mounted_dirs))))
-  $(warning Mounted EXTERNAL_MODULE_DIRS: $(_mounted_dirs))
-  $(error Mapping EXTERNAL_MODULE_DIRS in docker is not supported for directories with the same name)
+  $(warning Mounted EXTERNAL_MODULE_PATHS: $(_mounted_dirs))
+  $(error Mapping EXTERNAL_MODULE_PATHS in docker is not supported for directories with the same name)
 endif
 
 # Handle worktree by mounting the git common dir in the same location
```

I do not know why this makefile is included as I do not use Docker ?

I continue testing your PR, still having troubles but maybe on my side.

gdoffe commented:
a module cannot be renamed with "MODULE =" in Makefile, you confirm ?
(I'm fine with that ;) )

gdoffe commented on makefiles/dependency_resolution.inc.mk:
```diff
@@ -2,6 +2,12 @@
 # transient dependencies
 
+# Locate used modules in $(EXTERNAL_MODULE_DIRS). Internal modules cannot depend on external
+# ones (for obvious reasons) and dependencies for external modules are not supported. Hence,
```
> dependencies for external modules are not supported

What do you mean ? I have external modules depending on other external modules and it works fine with your PR.

maribu replied:
Good you noticed. That is a leftover. I initially wanted to keep them out of the recursive dependency resolution to avoid breakage. But then I checked and noticed that it eas previously supported, so I readded it. I just forgot to update the comment.

maribu replied:
Fixed

maribu commented:
> a module cannot be renamed with "MODULE =" in Makefile, you confirm ?
> (I'm fine with that ;) )

The current implementation locates them by name. It would be more involved to support mismatching folder and module names. I forgot that for internal modules this is supported, kind of. (For internal modules one has to manually extend `DIRS` if there is a mismatch.)

This limitation should definitely be documented. Thanks for pointing this out.

maribu commented:
> Just a point, do you agree that makefiles/docker.inc.mk should be patched too to replace EXTERNAL_MODULE_DIRS by EXTERNAL_MODULE_PATHS ?

I haven't looked into docker integration yet. But your suggestion sounds good. But let's wait for some feedback on the interface change first. If that is agreed upon, I can fix the docker integration.

gdoffe commented:
> I haven't looked into docker integration yet. But your suggestion sounds good. But let's wait for some feedback on the interface change first. If that is agreed upon, I can fix the docker integration.

The problem is I do neither use docker. ;)
But docker.inc.mk is included whether you build with docker or not:
https://github.com/RIOT-OS/RIOT/blob/master/Makefile.include#L162
So as my modules are outside of $(RIOTBASE) they are prepared to be mapped through a docker volume.
Thus the sanity check is failing due to using EXTERNAL_MODULE_DIRS instead of EXTERNAL_MODULE_PATHS.
So whether you use docker or not, your PR fails with external modules outside of $(RIOTBASE) due to always including docker.inc.mk.

However, with patched docker.inc.mk that works very well for me with external modules outside of $(RIOTBASE).
https://github.com/cogip/mcu-firmware/tree/gdo_test_makefiles

```bash
gdo@gdo-desktop:~/Developpement/Informatique/Personnel/cogip/mcu-firmware$ make -C examples/shell_menu/
make : on entre dans le répertoire « /home/gdo/Developpement/Informatique/Personnel/cogip/mcu-firmware/examples/shell_menu »
Building application "shell-menu" for "native" with MCU "native".

"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/boards/native
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/boards/native/drivers
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/core
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/cpu/native
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/cpu/native/periph
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/cpu/native/stdio_native
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/drivers
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/drivers/periph_common
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/auto_init
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/ps
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/shell
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/shell/commands
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/mcu-firmware/sys/shell_menu
   text	   data	    bss	    dec	    hex	filename
  38194	    776	  48092	  87062	  15416	/home/gdo/Developpement/Informatique/Personnel/cogip/mcu-firmware/examples/shell_menu/bin/native/shell-menu.elf
make : on quitte le répertoire « /home/gdo/Developpement/Informatique/Personnel/cogip/mcu-firmware/examples/shell_menu »
gdo@gdo-desktop:~/Developpement/Informatique/Personnel/cogip/mcu-firmware$ make -C examples/lds01/
make : on entre dans le répertoire « /home/gdo/Developpement/Informatique/Personnel/cogip/mcu-firmware/examples/lds01 »
Building application "lds01_example" for "nucleo-f446re" with MCU "stm32".

"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/boards/nucleo-f446re
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/boards/common/nucleo
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/core
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/cpu/stm32
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/cpu/cortexm_common
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/cpu/cortexm_common/periph
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/cpu/stm32/periph
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/cpu/stm32/stmclk
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/cpu/stm32/vectors
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/drivers
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/drivers/periph_common
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/auto_init
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/div
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/event
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/isrpipe
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/malloc_thread_safe
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/newlib_syscalls_default
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/pm_layered
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/shell
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/stdio_uart
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/tsrb
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/RIOT/sys/xtimer
"make" -C /home/gdo/Developpement/Informatique/Personnel/cogip/mcu-firmware/drivers/lds01
   text	   data	    bss	    dec	    hex	filename
  17928	    136	   7048	  25112	   6218	/home/gdo/Developpement/Informatique/Personnel/cogip/mcu-firmware/examples/lds01/bin/nucleo-f446re/lds01_example.elf
[##                                                ] 4% (18064 / 524288 Flash)
[                                                  ] 1% (7184 / 134217728 RAM)
make : on quitte le répertoire « /home/gdo/Developpement/Informatique/Personnel/cogip/mcu-firmware/examples/lds01 »
```

maribu commented:
> The problem is I do neither use docker. ;)
> But docker.inc.mk is included whether you build with docker or not:
> https://github.com/RIOT-OS/RIOT/blob/master/Makefile.include#L162
> So as my modules are outside of $(RIOTBASE) they are prepared to be mapped through a docker volume.

Interestingly, I couldn't reproduce the issue. I wonder if the issue was caused when using `EXTERNAL_MODULE_DIRS` with relative paths. I adapted `Makefile.include` to make them absolute, just as all other paths. Could you check if this fixes the issue?

gdoffe commented:
> Could you check if this fixes the issue?

Yes it is, even with DOCKER_BUILD=1.
Thank you for your help.

fjmolinas commented:
I like this PR a lot, it will need careful testing though, I'll take a closer look tomorrow. Regarding the semantics If I understand the same variable name is kept but it works differently? Will this work with setups using the previous workflow?

maribu commented:
> Regarding the semantics If I understand the same variable name is kept but it works differently?

Yes.

> Will this work with setups using the previous workflow?

No, it will break existing setups. It would like be possible to provide backward compatibility. I'm not sure if it is worth is. I don't expect many users of this features - it is more of a niche feature IMO. But all heavy users of this will likely like the new interface better and will be happy to migrate.

Maybe it is best to ask in the forum. At the very least, it will help creating attention and prevent users being surprised by the change.

fjmolinas commented on doc/doxygen/src/creating-modules.md:
```diff
@@ -68,13 +68,19 @@ their dependencies.
   API headers include paths to the `USEMODULE_INCLUDES` variable.
 * `Makefile.dep` file to set module dependencies
 
+***NOTE:*** The name of an external module must be unique (both in regard to other
+            external modules, as well to native RIOT modules). Additionally, the
+            directory containing the module must match the module name, e.g.
+            module `foo` must be located in `<PATH_IN_EXTERNAL_MODULE_DIRS>/foo`.
```
Having had a closer look at the PR I think this hard requirement is the main change that could potentially break external setups as well as make it harder to host bigger out-of-tree projects. I can think of any very good solutions though... @gdoffe used `sed` in his PR, but performing multiple shell calls is a bit slow.

At the same time `EXTERNAL_MODULE_DIRS` although less convenient actually maps what happens with Modules currently in RIOT regarding this, modules in RIOT are not automatically part of a search path either.

But this is only regarding compilation, `Makefile` inclusion inherently requires what is written in this `note` since it's based on rules such as: `-include $(USEPKG:%=$(RIOTPKG)/%/Makefile.dep)` or `-include $(sort $(USEMODULE:%=$(RIOTBASE)/drivers/%/Makefile.dep))` or `-include $(EXTERNAL_MODULE_DIRS:%=%/Makefile.dep)`, so if the module name does not match the directory name then no `Makefile.%` would get included. (Don't know if somewhere it says that a module with `Makefile.%` must have the same name as its directory for drivers).

So anyway we are already dealing with nondefault behavior to get some convenience, so trading one for another is fine if most agree, let's give you a forum post some time and then maybe add a `Make` warning for a while if there is `EXTERNAL_MODULE_DIRS` usage somewhere.

benpicco replied:
Aren't unique module names already a requirement for RIOT modules?

fjmolinas replied:
> Aren't unique module names already a requirement for RIOT modules?

For the same build yes, but if there is no overlap, lets say between two different BOARD's, then both can have a module names `periph`. But here the important part is that directories - module would be 1:1, and although in the same build module names must be unique directories don't.

maribu replied:
> can we have multiple EXTERNAL_MODULE_DIRS though?

Sure :-) the same already applies to EXTERNAL_BOARD_DIRS. I have one Repo with work boards, and one repo with home grown private boards in it :-)

benpicco commented:
> No, it will break existing setups. It would likely be possible to provide backward compatibility. I'm not sure if it is worth is. 

As long as it's failing loud and with an easy to understand error message, I see no issue.

> I don't expect many users of this features - it is more of a niche feature IMO. 

At least we use this feature a lot for our firmware. Handling custom modules the same way as RIOT modules is a pretty useful thing.

> But all heavy users of this will likely like the new interface better and will be happy to migrate.

You are right with that :D

fjmolinas commented:
>> I don't expect many users of this features - it is more of a niche feature IMO.

> At least we use this feature a lot for our firmware. Handling custom modules the same way as RIOT modules is a pretty useful thing.

And how do you see this change affecting your directory structure? Will it work well for you? I think a complex use case is the only thing we are really missing to move forward with this, someone that uses this heavily.

benpicco commented:
All the external modules are already in a single folder

Currently we have

```
MODULEDIR ?= $(abspath $(dir $(lastword $(MAKEFILE_LIST))))

KNOWN_EXTERNAL_MODULES = $(foreach md, $(dir $(wildcard $(MODULEDIR)/*/.)), $(shell basename $(md)))
EXTERNAL_MODULE_DIRS += $(addprefix $(MODULEDIR)/, $(filter $(KNOWN_EXTERNAL_MODULES), $(USEMODULE)))
```

to avoid having to manually add `KNOWN_EXTERNAL_MODULES` for each external module.

gdoffe commented:
On our side, for some modules, we have a sub-directory `shell_<module_name>`.
So each module having a shell subdirectory must be added to `EXTERNAL_MODULE_DIRS`.
We handle it this way:
```
MCUFW_MODULE_DIRS = " \                                                         
                    controllers \                                               
                    drivers \                                                   
                    lib \                                                       
                    planners \                                                  
                    platforms \                                                 
                    sys \                                                       
                    "                                                           
                                                                                
# All directories starting with 'shell_' are shell modules.                     
# List them to add their parent directory to EXTERNAL_MODULE_DIRS.              
shell_module_dirs = $(wildcard $(MCUFW_MODULE_DIRS:%=$(MCUFIRMWAREBASE)/%/*/shell_*/.))
EXTERNAL_MODULE_DIRS += $(MCUFW_MODULE_DIRS:%=$(MCUFIRMWAREBASE)/%/) $(shell_module_dirs:%=%/..)
```

maribu commented:
@gdoffe: Maybe the submodule approach used e.g. in `sys/fmt` with `fmt_table` being a submodule that adds `sys/fmt/table.c` to the list of sources to compile for `sys/fmt` would be useful for your use case? I can imagine that the shell command is a single C file anyway.

That being said, the layout of having related modules within the "base" module makes a lot of sense to me. And the submodule approach only works for tiny stuff that fits in a single C file. Would it make sense to support extending `EXTERNAL_MODULE_DIRS` via `+=` from within external modules `Makefile.dep`?

(Note that I do hope to rewrite the module location code in the build system to use the same implementation also for internal modules. E.g. `sys/net/gnrc` would also be something to add to `INTERNAL_MODULE_DIRS` if and only if `gnrc` is used. Hence, this mechanism would make sense later on anyway.)

fjmolinas commented:
> @gdoffe: Maybe the submodule approach used e.g. in `sys/fmt` with `fmt_table` being a submodule that adds `sys/fmt/table.c` to the list of sources to compile for `sys/fmt` would be useful for your use case? I can imagine that the shell command is a single C file anyway.
> 
> That being said, the layout of having related modules within the "base" module makes a lot of sense to me. And the submodule approach only works for tiny stuff that fits in a single C file. Would it make sense to support extending `EXTERNAL_MODULE_DIRS` via `+=` from within external modules `Makefile.dep`?
> 
> (Note that I do hope to rewrite the module location code in the build system to use the same implementation also for internal modules. E.g. `sys/net/gnrc` would also be something to add to `INTERNAL_MODULE_DIRS` if and only if `gnrc` is used. Hence, this mechanism would make sense later on anyway.)

Does it make sense to adapt the current test to handle this? Also regarding @gdoffe use case wouldn't this unique directory name requirement be an issue, they have a `sys` directory which already exists in RIOT, same for `drivers`.

maribu commented:
> Also regarding @gdoffe use case wouldn't this unique directory name requirement be an issue, they have a `sys` directory which already exists in RIOT, same for `drivers`.

As far as I understood, the `sys` directory is not a module in @gdoffe setup, but a search directory containing modules. For entries in `EXTERNAL_MODULE_DIRS` it is enough if the full absolute path is unique - which is always the case, as two identical absolute paths should identify the same directory.

fjmolinas commented:
> > Also regarding @gdoffe use case wouldn't this unique directory name requirement be an issue, they have a `sys` directory which already exists in RIOT, same for `drivers`.
> 
> As far as I understood, the `sys` directory is not a module in @gdoffe setup, but a search directory containing modules. For entries in `EXTERNAL_MODULE_DIRS` it is enough if the full absolute path is unique - which is always the case, as two identical absolute paths should identify the same directory.

Well they are using it 26a536d so must been it works for them :)

fjmolinas commented:
Since we are in soft-feature freeze let's keep this on hold, but I think we have enough input to move forward with this in the next release cycle. Otherwise, maybe @vincent-d @jue89 and @jeandudey might also have an opinion since they use/used RIOT in commercial products AFAIK.

jue89 commented:
Long story short: I'd love to have this PR in RIOT master. Currently, our Makefiles are filled with `EXTERNAL_MODULE_DIR += ...`. I'm happy to migrate this to just a single line. All our external modules are inside one folder.

jeandudey commented:
+1 on having this on master, for large projects I tend to have a `sys/` folder with all the modules and keeping it in sync with `USEMODULE` is a little bit problematic.

maribu commented:
I'll try to fix the remaining issues until hard feature freeze. So the very moment the release branch is detached form `master`, this should be ready to get in :-)

fjmolinas commented:
Hard feature freeze is in now, so this one can be pushed forward!

maribu commented:
Murdock is happy. Time to squash?

fjmolinas commented on makefiles/dependency_resolution.inc.mk:
```diff
@@ -2,6 +2,10 @@
 # until no new modules, pkgs, or features are pull in order to catch all
 # transient dependencies
 
+# Locate used modules in $(EXTERNAL_MODULE_DIRS).
+EXTERNAL_MODULE_PATHS := $(sort $(foreach dir,$(EXTERNAL_MODULE_DIRS),\
+  $(foreach mod,$(USEMODULE),$(dir $(wildcard $(dir)/$(mod)/Makefile)))))
+
```
Re-looking at the code, and comment/question in https://github.com/RIOT-OS/RIOT/pull/16104#issuecomment-806163303, this would mean all `EXTERNAL_MODULE_DIRS` would need to be absolute paths no?

maribu commented on Makefile.include:
```diff
@@ -121,6 +124,9 @@ override DLCACHE_DIR            := $(abspath $(DLCACHE_DIR))
 EXTERNAL_BOARD_DIRS             := $(foreach dir,\
                                              $(EXTERNAL_BOARD_DIRS),\
                                              $(abspath $(dir)))
+EXTERNAL_MODULE_DIRS            := $(foreach dir,\
+                                             $(EXTERNAL_MODULE_DIRS),\
+                                             $(abspath $(dir)))
```
Here any entries in `EXTERNAL_MODULE_DIRS` that are relative will be converted to absolute paths.

fjmolinas commented:
@maribu you can squash, I don't think I'm able to foresee all corner cases, so there might be issues along the way, but let's start using this to encounter them. I'll do some testing tomorrow, but otherwise lets go ahead.

fjmolinas reviewed (commented):
Some testing where I removed the `external_modules` directory in `tests/external_module_dirs` and put it outside of tree and exported the location:

<details> <summary> compile test <b>OK</b> </summary>

```
make -C tests/external_module_dirs/
Building application "tests_external_module_dirs" for "native" with MCU "native".

"make" -C /home/francisco/workspace/RIOT/boards/native
"make" -C /home/francisco/workspace/RIOT/boards/native/drivers
"make" -C /home/francisco/workspace/RIOT/core
"make" -C /home/francisco/workspace/RIOT/cpu/native
"make" -C /home/francisco/workspace/RIOT/cpu/native/periph
"make" -C /home/francisco/workspace/RIOT/cpu/native/stdio_native
"make" -C /home/francisco/workspace/RIOT/drivers
"make" -C /home/francisco/workspace/RIOT/drivers/periph_common
"make" -C /home/francisco/workspace/RIOT/sys
"make" -C /home/francisco/workspace/RIOT/sys/auto_init
"make" -C /home/francisco/workspace/RIOT/sys/luid
"make" -C /home/francisco/workspace/RIOT/sys/random
"make" -C /home/francisco/workspace/tmp/external_modules/external_module
   text	   data	    bss	    dec	    hex	filename
  29345	    628	  47820	  77793	  12fe1	/home/francisco/workspace/RIOT/tests/external_module_dirs/bin/native/tests_external_module_dirs.elf
```
</details>

<details> <summary> compile test in docker <b>OK</b> </summary>

```
Launching build container using image "riot/riotbuild:latest".
docker run --rm --tty --user $(id -u) -v '/usr/share/zoneinfo/Europe/Paris:/etc/localtime:ro' -v '/home/francisco/workspace/RIOT:/data/riotbuild/riotbase:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'   -v '/home/francisco/workspace/tmp/external_modules:/data/riotbuild/external/external_modules:delegated'      -w '/data/riotbuild/riotbase/tests/external_module_dirs/' 'riot/riotbuild:latest' make  'EXTERNAL_MODULE_DIRS=/data/riotbuild/external/external_modules /data/riotbuild/riotbase/tests/external_module_dirs/external_modules'    
Building application "tests_external_module_dirs" for "native" with MCU "native".

"make" -C /data/riotbuild/external/external_modules/external_module
"make" -C /data/riotbuild/riotbase/boards/native
"make" -C /data/riotbuild/riotbase/boards/native/drivers
"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/cpu/native
"make" -C /data/riotbuild/riotbase/cpu/native/periph
"make" -C /data/riotbuild/riotbase/cpu/native/stdio_native
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/sys/auto_init
"make" -C /data/riotbuild/riotbase/sys/luid
"make" -C /data/riotbuild/riotbase/sys/random
   text	   data	    bss	    dec	    hex	filename
  28792	    628	  47756	  77176	  12d78	/data/riotbuild/riotbase/tests/external_module_dirs/bin/native/tests_external_module_dirs.elf
```
</details>

<details> <summary> compile test with setup matching previous handling <b>FAIL</b> </summary>

```
BUILD_IN_DOCKER=1 make -C tests/external_module_dirs/
Launching build container using image "riot/riotbuild:latest".
docker run --rm --tty --user $(id -u) -v '/usr/share/zoneinfo/Europe/Paris:/etc/localtime:ro' -v '/home/francisco/workspace/RIOT:/data/riotbuild/riotbase:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'   -v '/home/francisco/workspace/tmp/external_module:/data/riotbuild/external/external_module:delegated'      -w '/data/riotbuild/riotbase/tests/external_module_dirs/' 'riot/riotbuild:latest' make  'EXTERNAL_MODULE_DIRS=/data/riotbuild/external/external_module /data/riotbuild/riotbase/tests/external_module_dirs/external_modules'    
Building application "tests_external_module_dirs" for "native" with MCU "native".

/data/riotbuild/riotbase/tests/external_module_dirs/main.c:23:10: fatal error: external_module.h: No such file or directory
 #include "external_module.h"
          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
/data/riotbuild/riotbase/Makefile.base:107: recipe for target '/data/riotbuild/riotbase/tests/external_module_dirs/bin/native/application_tests_external_module_dirs/main.o' failed
make[1]: *** [/data/riotbuild/riotbase/tests/external_module_dirs/bin/native/application_tests_external_module_dirs/main.o] Error 1
/data/riotbuild/riotbase/Makefile.include:641: recipe for target 'application_tests_external_module_dirs.module' failed
make: *** [application_tests_external_module_dirs.module] Error 2
make: *** [/home/francisco/workspace/RIOT/makefiles/docker.inc.mk:314: ..in-docker-container] Error 2
```
</details>

So the thing that remains IMO is how to notify external users of the change, since the variable name has not changed it's hard to add a warning. Too bad the current name is quite good already.... I would suggest adding a warning if `EXTERNAL_MODULE_DIRS` is set for a release or two, what do you think @maribu?

fjmolinas commented:
> Some testing where I removed the `external_modules` directory in `tests/external_module_dirs` and put it outside of tree and exported the location:
> compile test OK
> compile test in docker OK
> compile test with setup matching previous handling FAIL
> 
> So the thing that remains IMO is how to notify external users of the change, since the variable name has not changed it's hard to add a warning. Too bad the current name is quite good already.... I would suggest adding a warning if `EXTERNAL_MODULE_DIRS` is set for a release or two, what do you think @maribu?

```
ifneq (,$(EXTERNAL_BOARD_DIRS))
  $(info Warning! EXTERNAL_BOARD_DIRS is a serach folder since 2021.07-branch, see https://doc.riot-os.org/creating-modules.html)
endif
```

fjmolinas commented on Makefile.include:
```diff
@@ -96,16 +96,21 @@ ifneq ($(abspath $(RIOTBASE)/boards),$(abspath $(RIOTBOARD)))
   EXTERNAL_BOARD_DIRS += $(BOARDSDIR)
+
+  ifneq (,$(EXTERNAL_BOARD_DIRS))
+    $(info Warning! EXTERNAL_BOARD_DIRS is a serach folder since 2021.07-branch, see \
```
```suggestion
    $(info Warning! EXTERNAL_BOARD_DIRS is a search folder since 2021.07-branch, see \
```

ups I suggested a typo

fjmolinas replied:
Can you also add a comment on when this will be removed?

fjmolinas replied:
You can squash right away as well.

fjmolinas reviewed (approved):
First ACK! There is consensus among users that this is a more convenient solution, so lets go ahead with the change.

fjmolinas commented:
@maribu if this gets merged, please announce on that discourse post.

fjmolinas commented:
@jue89 @jeandudey @benpicco as users and maintainers I think your ACKS on this PR would be nice :)

jue89 commented:
I'll test this PR tonight with my application :)

benpicco reviewed (approved):
works like a charm!

All I had to do was

```patch
--- a/modules/external_module_dirs.mk
+++ b/modules/external_module_dirs.mk
@@ -1,4 +1 @@
-MODULEDIR ?= $(abspath $(dir $(lastword $(MAKEFILE_LIST))))
-
-KNOWN_EXTERNAL_MODULES = $(foreach md, $(dir $(wildcard $(MODULEDIR)/*/.)), $(shell basename $(md)))
-EXTERNAL_MODULE_DIRS += $(addprefix $(MODULEDIR)/, $(filter $(KNOWN_EXTERNAL_MODULES), $(USEMODULE)))
+EXTERNAL_MODULE_DIRS := $(dir $(lastword $(MAKEFILE_LIST)))
```

benpicco commented:
For some reason `tests/external_board_native` is failing on CI

fjmolinas commented on Makefile.include:
```diff
@@ -93,14 +97,22 @@ ifneq ($(abspath $(RIOTBASE)/boards),$(abspath $(RIOTBOARD)))
+
+    # API change warning for EXTERNAL_MODULE_DIRS, remove by 2021.10
+    ifneq (,$(EXTERNAL_MODULE_DIRS))
+      ifeq (1,$(WARNING_EXTERNAL_MODULE_DIRS))
```
Does this make sense now? I had added it so that a user that had migrated could disable the warning, but maybe it does not make sense.

maribu replied:
I think power users will be happy to disable the warning, once the have fully migrated :-)

fjmolinas replied:
Ahh, I was just confused because I didn't see the definition, but it's just above, isn't it better to keep it here with where its used (feel free to dismiss, just commenting since with a red Murdock this will require another build anyway)

jue89 reviewed (approved):
Sweet! Works perfectly with my application.

maribu commented:
> For some reason `tests/external_board_native` is failing on CI

The reason was stupid me introducing a bug while cleaning up the code :-/ I fixed the issue and squashed right away.

benpicco commented:
Somehow the static-tests script is not coming to a conclusion

jue89 commented:
> Somehow the static-tests script is not coming to a conclusion

It timed out . I can't see any reason for this. Maybe we just try to rerun them?

fjmolinas commented:
> > Somehow the static-tests script is not coming to a conclusion
> 
> It timed out . I can't see any reason for this. Maybe we just try to rerun them?

This is weird, it should take around 3-4mins, and its been going on for 10, there must be an issue.

maribu commented:
Looks like cppcheck is running in and endless loop now :/

maribu commented:
Locally, cppcheck terminates. Maybe I should check again using the riotbuild docker image.

fjmolinas commented:
> Locally, cppcheck terminates. Maybe I should check again using the riotbuild docker image.

I ran it in docker and it still terminates for me... @maribu can you force push a change in the commit message or something?

fjmolinas commented:
> Locally, cppcheck terminates. Maybe I should check again using the riotbuild docker image.

Actually, I think it's vera++ and not cppcheck.

fjmolinas commented:
@maribu can you try with this patch?

```diff
diff --git a/dist/tools/vera++/check.sh b/dist/tools/vera++/check.sh
index ce74912d48..61a5a7158b 100755
--- a/dist/tools/vera++/check.sh
+++ b/dist/tools/vera++/check.sh
@@ -16,7 +16,7 @@ CURDIR=$(cd "$(dirname "$0")" && pwd)
 # tests/pkg_utensor/models/deep_mlp_weight.hpp is an auto-generated file
 # with lots of commas so T009 takes very long. Since it is auto-generated, just
 # exclude it.
-EXCLUDE='^(.+/vendor/|dist/tools/coccinelle/include|dist/tools/fixdep/fixdep.c|dist/tools/lpc2k_pgm/src|tests/pkg_utensor/models)'
+EXCLUDE='^(.+/vendor/|dist/tools/coccinelle/include|dist/tools/fixdep/fixdep.c|dist/tools/lpc2k_pgm/src|tests/pkg_utensor/external_modules/models)'
 FILES=$(changed_files)

 if [ -z "${FILES}" ]; then
```

fjmolinas commented:
Gooo!

maribu commented:
Thanks everyone for the feedback and special thanks to @fjmolinas for figuring out the issue with vera++!

fjmolinas commented:
A bit too late but I realize this broke dependency handling for EXTERNAL_MODULE when using Kconfig since `EXTERNAL_MODULE_KCONFIG` is now using `EXTERNAL_MODULE_PATH` which is only set in normal dependency resolution.

fjmolinas commented:
> A bit too late but I realize this broke dependency handling for EXTERNAL_MODULE when using Kconfig since `EXTERNAL_MODULE_KCONFIG` is now using `EXTERNAL_MODULE_PATH` which is only set in normal dependency resolution.

I think I have the fix though.
